lib/errors/index.ts

```diff
@@ -1,89 +1,94 @@
-import errorsObj from './arsenalErrors';
-
-/**
- * ArsenalError
- *
- * @extends {Error}
- */
-class ArsenalError extends Error {
-    code: number
-    description: string
-
-    /**
-     * constructor.
-     *
-     * @param {string} type - Type of error or message
-     * @param {number} code - HTTP status code
-     * @param {string} desc - Verbose description of error
-     */
-    constructor(type, code, desc) {
-        super(type);
+import * as http from 'http';
+import * as rawErrors from './arsenalErrors';
+import * as types from './types';
+
+export * from './types';
+
+/** Mapping used to determine an error type. */
+export type Is = { [Name in types.Name]: boolean };
+/** Mapping of all possible Errors */
+export type Errors = { [Property in keyof types.Names]: ArsenalError };
+
+const entries = Object.entries(rawErrors);
 
-        /**
-         * HTTP status code of error
-         * @type {number}
-         */
-        this.code = code;
+// This contains some metaprog. Be careful.
+// Proxy can be found on MDN.
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy
+const createIs = (type: types.Name) => {
+    const get = (_: {}, value: string | symbol) => type === value;
+    return new Proxy({}, { get }) as Is;
+};
 
-        /**
-         * Description of error
-         * @type {string}
-         */
-        this.description = desc;
+export class ArsenalError extends Error {
+    /** HTTP status code. Example: 401, 403, 500, ... */
+    #code: number;
+    /** Text description of the error. */
+    #description: string;
+    /** Type of the error. Belongs to errors.types. */
+    #type: types.Name;
+    /** Object used to determine the error type.
+     * Example: error.is.InternalError */
```

Comment:
what is the benefit of this approach, vs a simpler `error.is(InternalError)` ?

Granted, it looks similar to the previous code ; but since we need to change it anyway, we may as well use simpler approach (and avoid the meta-programming)...

Reply:
This is mainly in comparison with the two ways of writing. Today, we are doing this:

```javascript
function example() {
  myfun(function(err, res) {
    if (err && err.InternalError) {
      // Do some stuff with error.
    } else {
      // Do some stuff with the res.
    }
  });
}
```

The TS equivalent we can do, with proper type-checking is the proposed solution:

```typescript
// Here the export default is there, because we can’t do another way.
function example() {
  myfun(function(err, res) {
    if (err?.is?.InternalError) {
      // Do some stuff with error.
    } else {
      // Do some stuff with the res.
    }
  });
}
```

The advantage here is the low code change, and the fluent API. The object is self-discoverable, you can easily understand what it is even if you are not so much aware of the package. At first we tried a simpler solution like you described, but the best we can do:

```javascript
import errors from 'errors';

function example() {
  myfun(function(err, res) {
    if (err?.is?.(errors.InternalError)) {
      // Do some stuff with error.
    } else {
      // Do some stuff with the res.
    }
  });
}
```

Besides the syntax for conditional function call (`object?.()`) which feels a little weird, we are forced to import the package errors and to have a function call, which feels less native than just accessing a property of the error. Int the end the difference is not so important, and can still benefit from the type-checking in both case, with the same type casts. It's more a matter of taste and style-guides/guidelines.

On the specific meta-programming part, `Proxy` are a commonly used pattern in such cases, but using meta-prog really depends on teams, so I don't really have an opinion on whether we should keep it or get rid of it. We mainly wanted to show the integration of meta-prog with TypeScript (because in some cases meta-prog is mandatory) and how we can leverage power of modern JS while having TS safety all the way through.